README.md
# djot.zig

From <https://djot.net/>:

> Djot is a light markup syntax. It derives most of its features from commonmark,
but it fixes a few things that make commonmark's syntax complex and difficult to
parse efficiently. It is also much fuller-featured than commonmark, with support
for definition lists, footnotes, tables, several new kinds of inline formatting
(insert, delete, highlight, superscript, subscript), math, smart punctuation,
attributes that can be applied to any element, and generic containers for
block-level, inline-level, and raw content.

"djot.zig" is a library for parsing djot files and converting them to HTML. It
also includes a command line interface for converting djot to HTML.

## State of djot.zig

djot.zig is very much a work in progress. There are many features to implement,
and the API will almost certainly change. If you want a fully featured version
of djot, consider using the [official version implemented in Lua](https://github.com/jgm/djot).

### Tests

djot.zig uses the HTML test cases from the [official version implemented in Lua](https://github.com/jgm/djot/tree/47f9b3b3db91985180603ca5263ea2ec83d3e75d/test),
and a couple other test cases taken from the [syntax description](https://htmlpreview.github.io/?https://github.com/jgm/djot/blob/master/doc/syntax.html).

Test results as of 2022-08-19:

```
$ zig build -fno-stage1 test
70 passed; 0 skipped; 165 failed.
```

That's 70/235 test cases passing; or about 29% of the tests passing.

### Inline Syntax

| Feature | Implemented? |
|---------|-------|
| inline links | partially |
| reference links | no |
| images | no |
| autolinks | yes |
| inline verbatim | yes |
| inline verbatim space removal | no |
| emphasis/strong | yes |
| inline highlight | no |
| super/subscript | no |
| insert/delete | no |
| smart punctuation | no |
| inline math | no |
| footnote references | no |
| hard line break | no |
| comments | no |
| emoji aliases | no |
| raw inline | no |
| spans | no |
| inline attributes | no |

### Block syntax

| Feature | Implemented? |
|---------|-------|
| paragraphs | yes |
| block quotes | yes |
| lists | partially |
| code blocks | no |
| thematic breaks | yes|
| raw blocks | no |
| div blocks | no |
| pipe tables | no |
| reference link definitions | no |
| footnote definitions | no |
| block attributes | no |
| implicit heading reference links | no |

## How to use

### Command Line Interface

Requires zig >= `0.10.0-dev.3567+95573dbee`.

```
$ git clone https://github.com/leroycep/djot.zig
$ cd djot.zig
$ zig build install -fno-stage1
$ echo "*Hello, world!*" | ./zig-out/bin/djot.zig
<p><strong>Hello, world!*"</strong></p>
```

### Using `djot.zig` as a Library

Requires zig stage2 >= `0.10.0-dev.3567+95573dbee`.

Let's make an application that extracts djot links from stdin! To get started
we'll create a new zig project using `zig init-exe` and cloning `djot.zig`:

```bash
$ mkdir new-project
$ cd new-project
$ zig init-exe

$ git clone https://github.com/leroycep/djot.zig
```

Now let's add `djot.zig` as a package:

```diff
--- old-build.zig
+++ build.zig
@@ -15,6 +15,7 @@
     exe.setTarget(target);
     exe.setBuildMode(mode);
     exe.install();
+    exe.addPackagePath("djot", "./djot.zig/src/djot.zig");
 
     const run_cmd = exe.run();
     run_cmd.step.dependOn(b.getInstallStep());
```

And replace the auto-generated main function for one that extracts links from
djot markup:

```diff
--- src/old-main.zig	2022-08-19 20:06:28.272239304 -0600
+++ src/main.zig	2022-08-19 20:23:15.261355146 -0600
@@ -1,19 +1,30 @@
 const std = @import("std");
+const djot = @import("djot");
 
+/// Extract all the links from djot markup that is passed to stdin
 pub fn main() !void {
-    // Prints to stderr (it's a shortcut based on `std.io.getStdErr()`)
-    std.debug.print("All your {s} are belong to us.\n", .{"codebase"});
+    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
+    defer _ = gpa.deinit();
 
-    // stdout is for the actual output of your application, for example if you
-    // are implementing gzip, then only the compressed bytes should be sent to
-    // stdout, not any debugging messages.
-    const stdout_file = std.io.getStdOut().writer();
-    var bw = std.io.bufferedWriter(stdout_file);
-    const stdout = bw.writer();
+    const stdin = std.io.getStdIn();
+    const source = try stdin.readToEndAlloc(gpa.allocator(), 50 * 1024 * 1024);
+    defer gpa.allocator().free(source);
 
-    try stdout.print("Run `zig build test` to run the tests.\n", .{});
+    var document = try djot.parse(gpa.allocator(), source);
+    defer document.deinit(gpa.allocator());
 
-    try bw.flush(); // don't forget to flush!
+    const stdout = std.io.getStdOut();
+    for (document.events.items(.tag)) |event_tag, event_index| {
+        switch (event_tag) {
+            .autolink,
+            .start_link,
+            .start_image_link,
+            => {
+                try stdout.writer().print("- {s}\n", .{document.asText(event_index)});
+            },
+            else => {},
+        }
+    }
 }
 
 test "simple test" {
```

Now let's test it out:

```bash
$ echo "fav websites: <https://ziglang.org>, <https://djot.net>" | zig build -fno-stage1 run
- https://ziglang.org
- https://djot.net
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Planned]

Partially done:

- inline links
- lists

Not started:

- reference links
- images
- inline verbatim space removal
- inline highlight
- super/subscript
- insert/delete
- smart punctuation
- inline math
- footnote references
- hard line break
- comments
- emoji aliases
- raw inline
- spans
- inline attributes
- code blocks
- raw blocks
- div blocks
- pipe tables
- reference link definitions
- footnote definitions
- block attributes
- implicit heading reference links
- C API


## [Unreleased]

### Added

- cli: Convert `djot` from stdin to `html`, `events`, or `tokens`
- autolinks
- inline verbatim
- inline emphasis/strong
- paragraph blocks
- quote blocks
- thematic breaks



tools/fancy-test-results.zig
const std = @import("std");
const bolt = @import("bolt");
const html = @import("html");

pub fn main() !void {
    // 1. get arguments for running test command
    // 2. run command and get stderr
    // 3. split stderr by test and final test results
    //   - each test starts with regex: `^\d+/\d+` (example: `235/235 test.html.verbatim 6... OK`)
    //   - tests results: `77 passed; 0 skipped; 158 failed.`
    // 4. Generate an html file and put it in "zig-out/test-results.html"

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();

    const args = try std.process.argsAlloc(gpa.allocator());
    defer std.process.argsFree(gpa.allocator(), args);

    const res = try std.ChildProcess.exec(.{
        .allocator = gpa.allocator(),
        .argv = args[1..],
        .max_output_bytes = 50 * 1024 * 1024,
    });
    defer {
        gpa.allocator().free(res.stdout);
        gpa.allocator().free(res.stderr);
    }

    const test_cases = try parseTestCases(gpa.allocator(), res.stderr);
    defer {
        gpa.allocator().free(test_cases.passed);
        gpa.allocator().free(test_cases.skipped);
        gpa.allocator().free(test_cases.failed);
    }

    var html_contents = std.ArrayList(u8).init(gpa.allocator());
    defer html_contents.deinit();

    const out = html_contents.writer();

    try out.print("<h2>Pass/Fail Total</h2>\n", .{});
    try out.print("<table>\n", .{});
    try out.print("<tr><td>{s}</td><td>{}</td></tr>\n", .{ "passed", test_cases.total.passed });
    try out.print("<tr><td>{s}</td><td>{}</td></tr>\n", .{ "skipped", test_cases.total.skipped });
    try out.print("<tr><td>{s}</td><td>{}</td></tr>\n", .{ "failed", test_cases.total.failed });
    try out.print("</table>\n", .{});

    try out.print("<h2>Failing Test Cases</h2>\n", .{});
    for (test_cases.failed) |test_case| {
        try out.print("<h3 id=\"", .{});
        try writeNameAsId(test_case.name, out);
        try out.print("\"><a href=\"#", .{});
        try writeNameAsId(test_case.name, out);
        try out.print("\">{}. {s}</a></h3>\n", .{ test_case.number, html.fmtEscapes(test_case.name) });
        try out.print("<pre><code>{}</code></pre>\n", .{html.fmtEscapes(test_case.result.err)});
    }

    try out.print("<h2>Skipped Test Cases</h2>\n", .{});
    try out.print("<ul>\n", .{});
    for (test_cases.skipped) |test_case| {
        try out.print("<li>{}. {s}</li>\n", .{ test_case.number, html.fmtEscapes(test_case.name) });
    }
    try out.print("</ul>\n", .{});

    try out.print("<h2>Passing Test Cases</h2>\n", .{});
    try out.print("<ul>\n", .{});
    for (test_cases.passed) |test_case| {
        try out.print("<li>{}. {s}</li>\n", .{ test_case.number, html.fmtEscapes(test_case.name) });
    }
    try out.print("</ul>\n", .{});

    try std.fs.cwd().writeFile("zig-out/test-results.html", html_contents.items);

    const path = try std.fs.realpathAlloc(gpa.allocator(), "zig-out/test-results.html");
    defer gpa.allocator().free(path);
    std.log.info("Wrote test results to file://{s}", .{path});
}

pub fn writeNameAsId(name: []const u8, writer: anytype) !void {
    for (name) |c| {
        switch (c) {
            'A'...'Z', 'a'...'z', '0'...'9', '_', '-' => try writer.writeByte(c),

            '.',
            ' ',
            => try writer.writeByte('-'),

            else => {},
        }
    }
}

const TestCases = struct {
    total: PassFailTotal,
    passed: []const TestCase,
    skipped: []const TestCase,
    failed: []const TestCase,
};

fn parseTestCases(allocator: std.mem.Allocator, source: []const u8) !TestCases {
    var passed = std.ArrayList(TestCase).init(allocator);
    defer passed.deinit();

    var skipped = std.ArrayList(TestCase).init(allocator);
    defer skipped.deinit();

    var failed = std.ArrayList(TestCase).init(allocator);
    defer failed.deinit();

    var cursor = Read{ .source = source, .index = 0 };

    while (true) {
        if (parsePassFailTotals(&cursor)) |total| {
            return TestCases{
                .total = total,
                .passed = passed.toOwnedSlice(),
                .skipped = skipped.toOwnedSlice(),
                .failed = failed.toOwnedSlice(),
            };
        }
        if (parseTestCase(&cursor)) |test_case| {
            switch (test_case.result) {
                .ok => try passed.append(test_case),
                .skip => try skipped.append(test_case),
                .err => try failed.append(test_case),
            }
            continue;
        }
        std.debug.print("\nHello: {s}\n", .{cursor.source[cursor.index..]});
        return error.Unknown;
    }
}

const Read = bolt.Read(u8, usize);

const TestCase = struct {
    number: u64,
    name: []const u8,
    result: Result,

    const Result = union(enum) {
        ok,
        skip,
        err: []const u8,
    };
};

fn parseTestCase(parent: *Read) ?TestCase {
    var cursor = parent.*;

    var header = parseTestCaseHeader(&cursor) orelse return null;
    var result_start = cursor.index;

    if (cursor.expectString("OK\n")) |_| {
        parent.* = cursor;
        return TestCase{
            .number = header.number,
            .name = header.name,
            .result = .ok,
        };
    } else if (cursor.expectString("SKIP\n")) |_| {
        parent.* = cursor;
        return TestCase{
            .number = header.number,
            .name = header.name,
            .result = .skip,
        };
    }

    while (cursor.next()) |_| {
        if (cursor.expect('\n')) |_| {
            var lookahead = cursor;
            if (parseTestCaseHeader(&lookahead)) |_| {
                break;
            }
            if (parsePassFailTotals(&lookahead)) |_| {
                break;
            }
        }
    }
    const result_end = cursor.index;

    parent.* = cursor;
    return TestCase{
        .number = header.number,
        .name = header.name,
        .result = .{ .err = cursor.source[result_start..result_end] },
    };
}

const TestCaseHeader = struct {
    number: u64,
    name: []const u8,
};

fn parseTestCaseHeader(parent: *Read) ?TestCaseHeader {
    var cursor = parent.*;

    const number = parseNumber(&cursor) orelse return null;
    _ = cursor.expect('/') orelse return null;
    _ = parseNumber(&cursor) orelse return null;
    _ = cursor.expect(' ') orelse return null;

    const name_start = cursor.index;
    var name_end = cursor.index;
    while (cursor.index < cursor.source.len) : (cursor.index += 1) {
        if (cursor.expectString("... ")) |_| {
            break;
        }
        name_end = cursor.index + 1;
    }

    parent.* = cursor;
    return TestCaseHeader{
        .number = number,
        .name = cursor.source[name_start..name_end],
    };
}

const PassFailTotal = struct {
    passed: u64,
    skipped: u64,
    failed: u64,
};

fn parsePassFailTotals(parent: *Read) ?PassFailTotal {
    var cursor = parent.*;

    const passed = parseNumber(&cursor) orelse return null;
    _ = cursor.expectString(" passed; ") orelse return null;

    const skipped = parseNumber(&cursor) orelse return null;
    _ = cursor.expectString(" skipped; ") orelse return null;

    const failed = parseNumber(&cursor) orelse return null;
    _ = cursor.expectString(" failed.") orelse return null;

    parent.* = cursor;
    return PassFailTotal{
        .passed = passed,
        .skipped = skipped,
        .failed = failed,
    };
}

fn parseNumber(parent: *Read) ?u64 {
    var cursor = parent.*;

    const start = cursor.expectInRange('0', '9') orelse return null;
    while (cursor.expectInRange('0', '9')) |_| {}

    const text = parent.source[start..cursor.index];
    const number = std.fmt.parseInt(u64, text, 10) catch return null;

    parent.* = cursor;
    return number;
}

test {
    var cursor = Read{
        .source = "77 passed; 0 skipped; 158 failed.",
        .index = 0,
    };
    try std.testing.expectEqual(@as(?PassFailTotal, PassFailTotal{
        .passed = 77,
        .skipped = 0,
        .failed = 158,
    }), parsePassFailTotals(&cursor));
}

test {
    var cursor = Read{
        .source = 
        \\235/235 test.html.verbatim 6... OK
        \\77 passed; 0 skipped; 158 failed.
        ,
        .index = 0,
    };

    const case = parseTestCase(&cursor) orelse return error.TestExpectedNotNull;
    try std.testing.expectEqual(@as(u64, 235), case.number);
    try std.testing.expectEqualStrings("test.html.verbatim 6", case.name);
    try std.testing.expectEqual(TestCase.Result.ok, case.result);

    try std.testing.expectEqual(@as(?TestCase, null), parseTestCase(&cursor));

    try std.testing.expectEqual(@as(?PassFailTotal, PassFailTotal{
        .passed = 77,
        .skipped = 0,
        .failed = 158,
    }), parsePassFailTotals(&cursor));
}
